x-pack/plugin/transform/src/main/java/org/elasticsearch/xpack/transform/transforms/pivot/CompositeBucketsChangeCollector.java

```diff
@@ -0,0 +1,350 @@
+        this.compositeAggregation = compositeAggregation;
+        this.fieldCollectors = fieldCollectors;
+    }
+
+    @Override
+    public boolean processSearchResponse(final SearchResponse searchResponse) {
+        final Aggregations aggregations = searchResponse.getAggregations();
+        if (aggregations == null) {
+            return true;
+        }
+
+        final CompositeAggregation agg = aggregations.get(compositeAggregation.getName());
+
+        Collection<? extends Bucket> buckets = agg.getBuckets();
+        afterKey = agg.afterKey();
+
+        if (buckets.isEmpty()) {
+            return true;
+        }
+
+        for (FieldCollector fieldCollector : fieldCollectors.values()) {
+            fieldCollector.collectChanges(buckets);
+        }
+
+        return false;
+    }
+
+    @Override
+    public QueryBuilder buildFilterQuery(long lastCheckpointTimestamp, long nextcheckpointTimestamp) {
+        BoolQueryBuilder filteredQuery = new BoolQueryBuilder();
+
+        for (FieldCollector fieldCollector : fieldCollectors.values()) {
+            QueryBuilder filter = fieldCollector.filterByChanges(lastCheckpointTimestamp, nextcheckpointTimestamp);
+            if (filter != null) {
+                filteredQuery.filter(filter);
+            }
+        }
+
+        return filteredQuery;
+    }
+
+    @Override
+    public SearchSourceBuilder buildChangesQuery(SearchSourceBuilder sourceBuilder, Map<String, Object> position, int pageSize) {
+
+        CompositeAggregationBuilder changesAgg = this.compositeAggregation;
+        changesAgg.size(pageSize).aggregateAfter(position);
+        sourceBuilder.aggregation(changesAgg);
+        sourceBuilder.size(0);
+        for (FieldCollector fieldCollector : fieldCollectors.values()) {
+            AggregationBuilder aggregationForField = fieldCollector.aggregateChanges();
+
+            if (aggregationForField != null) {
+                sourceBuilder.aggregation(aggregationForField);
+            }
+        }
+
+        return sourceBuilder;
+    }
+
+    @Override
+    public void clear() {
+        fieldCollectors.forEach((k, c) -> c.clear());
+    }
+
+    @Override
+    public Map<String, Object> getBucketPosition() {
+        return afterKey;
+    }
+
+    public static ChangeCollector buildChangeCollector(
+        CompositeAggregationBuilder compositeAggregationBuilder,
+        Map<String, SingleGroupSource> groups,
+        String synchronizationField
+    ) {
+        Map<String, FieldCollector> fieldCollectors = createFieldCollectors(groups, synchronizationField);
+        return new CompositeBucketsChangeCollector(compositeAggregationBuilder, fieldCollectors);
+    }
+
+    static Map<String, FieldCollector> createFieldCollectors(Map<String, SingleGroupSource> groups, String synchronizationField) {
+        Map<String, FieldCollector> fieldCollectors = new HashMap<>();
+
+        for (Entry<String, SingleGroupSource> entry : groups.entrySet()) {
```

Comment:
I don't know what we are gaining by not having `SingleGroupSource` have a `createFieldCollector` method. We would avoid this switch statement. 

Are we just trying to keep from adding classes to `core`?

Reply:
To me it seems unavoidable, config and implementation should be separate, like `PivotConfig` and `Pivot`. so far it was simple enough to put in into `...GroupSource` (which was totally ok). With the re-factoring it gets more complicated and there are already plans, that will add more code (maybe one day we end up with 1 class per `GroupSource`).

Yes keeping code away from `core` is a reason. If I would add it the implementation to `...GroupSource`, I would not only need to move the code there, but also the collector interface.

I think this change is similar to what is done in other places. The boiler-plate (switch) isn't great, but I do not see a need for something registry-style (like aggregations) at least at the moment.